### PR TITLE
[TD-4352]compact tsdb meta data feature

### PR DESCRIPTION
**Pull request by lichuang**

### CODE REVIEW

File: src/tsdb/src/tsdbCommit.c [hunk 1]
```diff
@@ -442,6 +452,114 @@ static int tsdbDropMetaRecord(STsdbFS *pfs, SMFile *pMFile, uint64_t uid) {
...
+
+    if (tsdbUpdateMetaRecord(pfs, &mf, pRecord->uid, pBuf, (int)pRecord->size, false) < 0) {
+      tsdbError("vgId:%d failed to update META record, uid %" PRIu64 " since %s", REPO_ID(pRepo), pRecord->uid,
+                tstrerror(terrno));
+      break;
+    }
+
+    pRecord = taosHashIterate(pfs->metaCache, pRecord);
+  }
+  code = 0;
```
Comment (kailixu) [inline src/tsdb/src/tsdbCommit.c:563]: while循环遇到异常break出来，也会执行到code=0这里，逻辑上你确认下是否正确？

Comment (lichuang) [reply src/tsdb/src/tsdbCommit.c:563]: 已修改

File: src/tsdb/src/tsdbMeta.c [hunk 2]
```diff
@@ -734,10 +746,7 @@ void tsdbUpdateTableSchema(STsdbRepo *pRepo, STable *pTable, STSchema *pSchema,
   TSDB_WUNLOCK_TABLE(pCTable);
 
   if (insertAct) {
-    int   tlen = tsdbGetTableEncodeSize(TSDB_UPDATE_META, pCTable);
-    void *buf = tsdbAllocBytes(pRepo, tlen);
-    ASSERT(buf != NULL);
-    tsdbInsertTableAct(pRepo, TSDB_UPDATE_META, buf, pCTable);
+    ASSERT(tsdbInsertNewTableAction(pRepo, pCTable) == 0);
```
Comment (cosmic-jc) [inline src/tsdb/src/tsdbMeta.c]: if NDEBUG is not defined, ASSERT is define as null. Please make sure the behaviour

Comment (cosmic-jc) [reply src/tsdb/src/tsdbMeta.c]: Besides, memory allocation error is possible.

File: src/tsdb/src/tsdbCommit.c [hunk 3]
```diff
@@ -442,6 +454,115 @@ static int tsdbDropMetaRecord(STsdbFS *pfs, SMFile *pMFile, uint64_t uid) {
...
+                tstrerror(terrno));
+      goto _err;
+    }
+
+    pRecord = taosHashIterate(pfs->metaCache, pRecord);
+  }
+  code = 0;
+
+_err:
+  TSDB_FILE_FSYNC(&mf);
```
Comment (cosmic-jc) [inline src/tsdb/src/tsdbCommit.c]: the calling of fsync is redundant when error occurs

File: src/tsdb/src/tsdbCommit.c [hunk 4]
```diff
@@ -442,6 +454,115 @@ static int tsdbDropMetaRecord(STsdbFS *pfs, SMFile *pMFile, uint64_t uid) {
...
+  while (pRecord) {
+    if (tsdbSeekMFile(pMFile, pRecord->offset + sizeof(SKVRecord), SEEK_SET) < 0) {
+      tsdbError("vgId:%d failed to seek file %s since %s", REPO_ID(pRepo), TSDB_FILE_FULL_NAME(pMFile),
+                tstrerror(terrno));
+      goto _err;
+    }
+    if (pRecord->size > maxBufSize) {
+      maxBufSize = pRecord->size;
+      void* tmp = realloc(pBuf, (size_t)maxBufSize);
+      if (tmp == NULL) {
```
Comment (kailixu) [inline src/tsdb/src/tsdbCommit.c:510]: 近期刚针对realloc分配失败进行了防止泄漏的处理，严谨些可以free(pBuf)

**Review by kailixu — APPROVED** [review]

File: src/tsdb/src/tsdbMeta.c (hunk 2 above)
Comment (lichuang) [reply src/tsdb/src/tsdbMeta.c]: fixed

File: src/tsdb/src/tsdbCommit.c (hunk 3 above)
Comment (lichuang) [reply src/tsdb/src/tsdbCommit.c]: fixed

**Review by cosmic-jc — APPROVED** [review]

File: src/tsdb/inc/tsdbFile.h [hunk 5]
```diff
@@ -38,7 +38,7 @@
 #define TSDB_FILE_IS_OK(tf) (TSDB_FILE_STATE(tf) == TSDB_FILE_STATE_OK)
 #define TSDB_FILE_IS_BAD(tf) (TSDB_FILE_STATE(tf) == TSDB_FILE_STATE_BAD)
 
-typedef enum { TSDB_FILE_HEAD = 0, TSDB_FILE_DATA, TSDB_FILE_LAST, TSDB_FILE_MAX, TSDB_FILE_META } TSDB_FILE_T;
```
Comment (hzcheng) [inline src/tsdb/inc/tsdbFile.h]: The file TSDB_FILE_META_TMP is not neccessary. You can use a new versioned meta file, For example: meta-ver1. This trick is used in TSDB file sync now.

Comment (cosmic-jc) [reply src/tsdb/inc/tsdbFile.h]: Great. Then there will be less one conflict for me!

Comment (lichuang) [reply src/tsdb/inc/tsdbFile.h]: thanks, fixed it

**Review by hzcheng — APPROVED** [review]

**Pull request merged**